ext/random/random.c

```diff
@@ -0,0 +1,1785 @@
+static zend_object_handlers random_engine_secure_object_handlers;
+static zend_object_handlers random_randomizer_object_handlers;
+
+static uint32_t rand_range32(const php_random_engine_algo *algo, void *state, uint32_t umax, bool *engine_unsafe) {
+	uint32_t result, limit;
+	size_t generated_size;
+
+	RANDOM_ENGINE_GENERATE_GUARD(algo, state, result, generated_size, engine_unsafe);
+	while (generated_size < sizeof(uint32_t)) {
+		uint32_t ret = 0;
+		size_t generate_size;
+
+		RANDOM_ENGINE_GENERATE_GUARD(algo, state, ret, generate_size, engine_unsafe);
+		result = (result << generate_size) | ret;
+
+		generated_size += generate_size;
+	}
+
+	/* Special case where no modulus is required */
+	if (UNEXPECTED(umax == UINT32_MAX)) {
+		return result;
+	}
+
+	/* Increment the max so the range is inclusive of max */
+	umax++;
+
+	/* Powers of two are not biased */
+	if ((umax & (umax - 1)) == 0) {
+		return result & (umax - 1);
+	}
+
+	/* Ceiling under which UINT32_MAX % max == 0 */
+	limit = UINT32_MAX - (UINT32_MAX % umax) - 1;
+
+	/* Discard numbers over the limit to avoid modulo bias */
+	while (UNEXPECTED(result > limit)) {
+		RANDOM_ENGINE_GENERATE_GUARD(algo, state, result, generated_size, engine_unsafe);
+		while (generated_size < sizeof(uint32_t)) {
+			uint32_t ret;
+			size_t generate_size;
+
+			RANDOM_ENGINE_GENERATE_GUARD(algo, state, ret, generate_size, engine_unsafe);
+
+			result = (result << generate_size) | ret;
+
+			generated_size += generate_size;
+		}
+	}
+
+	return result % umax;
+}
+
+#if ZEND_ULONG_MAX > UINT32_MAX
+static uint64_t rand_range64(const php_random_engine_algo *algo, void *state, uint64_t umax, bool *engine_unsafe) {
+	uint64_t result, limit;
+	size_t generated_size;
+
+	RANDOM_ENGINE_GENERATE_GUARD(algo, state, result, generated_size, engine_unsafe);
+
+	while (generated_size < sizeof(uint64_t)) {
+		uint64_t ret = 0;
+		size_t generate_size = 0;
+
+		RANDOM_ENGINE_GENERATE_GUARD(algo, state, ret, generate_size, engine_unsafe);
+
+		result = (result << generate_size) | ret;
+
+		generated_size += generate_size;
+	}
+
+	/* Special case where no modulus is required */
+	if (UNEXPECTED(umax == UINT64_MAX)) {
+		return result;
+	}
+
+	/* Increment the max so the range is inclusive of max */
+	umax++;
+
+	/* Powers of two are not biased */
+	if ((umax & (umax - 1)) == 0) {
+		return result & (umax - 1);
+	}
+
+	/* Ceiling under which UINT64_MAX % max == 0 */
+	limit = UINT64_MAX - (UINT64_MAX % umax) - 1;
+
+	/* Discard numbers over the limit to avoid modulo bias */
+	while (UNEXPECTED(result > limit)) {
+		RANDOM_ENGINE_GENERATE_GUARD(algo, state, result, generated_size, engine_unsafe);
+
+		while (generated_size < sizeof(uint64_t)) {
+			uint64_t ret = 0;
+			size_t generate_size = 0;
+
+			RANDOM_ENGINE_GENERATE_GUARD(algo, state, ret, generate_size, engine_unsafe);
+
+			result = (result << generate_size) | ret;
+
+			generated_size += generate_size;
+		}
+	}
```

Comment:
I managed to break this with a horrible userland RNG:

```
<?php

use Random\Engine;
use Random\Randomizer;

class MyEngine implements Engine {
    public function generate(): string {
      return "\xff\xff\xff\xff\xff\xff\xff\xff";
    }
}

$r = new Randomizer(new MyEngine());
var_dump($r->getInt(0, 500));
```

results in an endless loop. There should likely be a loop counter that throws an exception if a suitable number cannot be generated in 50 tries or so.

Reply:
fixed https://github.com/php/php-src/commit/ce5920ec5ddf01550d4bed61df8898db7758ce8a